src/sota_tools/ostree_http_repo_test.cc

```diff
@@ -0,0 +1,108 @@
+  EXPECT_FALSE(src_repo->LooksValid());
+}
+
+TEST(http_repo, getRef) {
+  TreehubServer server;
+  server.root_url("http://localhost:" + port);
+  OSTreeRepo::ptr src_repo = std::make_shared<OSTreeHttpRepo>(&server);
+  EXPECT_EQ(src_repo->GetRef("master").GetHash().string(),
+            std::string("16ef2f2629dc9263fdf3c0f032563a2d757623bbc11cf99df25c3c3f258dccbe"));
+}
+
+TEST(http_repo, GetObject) {
+  TreehubServer server;
+  server.root_url("http://localhost:" + port);
+  OSTreeRepo::ptr src_repo = std::make_shared<OSTreeHttpRepo>(&server);
+  const uint8_t hash[32] = {0x2a, 0x28, 0xda, 0xc4, 0x2b, 0x76, 0xc2, 0x01, 0x5e, 0xe3, 0xc4,
+                            0x1c, 0xc4, 0x18, 0x3b, 0xb8, 0xb5, 0xc7, 0x90, 0xfd, 0x21, 0xfa,
+                            0x5c, 0xfa, 0x08, 0x02, 0xc6, 0xe1, 0x1f, 0xd0, 0xed, 0xbe};
+  auto object = src_repo->GetObject(hash);
+  std::stringstream s;
+  s << object;
+  EXPECT_EQ(s.str(), std::string("2a/28dac42b76c2015ee3c41cc4183bb8b5c790fd21fa5cfa0802c6e11fd0edbe.dirmeta"));
+}
+
+TEST(http_repo, GetWrongObject) {
+  TreehubServer server;
+  server.root_url("http://localhost:" + port);
+  OSTreeRepo::ptr src_repo = std::make_shared<OSTreeHttpRepo>(&server);
+  const uint8_t hash[32] = {0x00, 0x28, 0xda, 0xc4, 0x2b, 0x76, 0xc2, 0x01, 0x5e, 0xe3, 0xc4,
+                            0x1c, 0xc4, 0x18, 0x3b, 0xb8, 0xb5, 0xc7, 0x90, 0xfd, 0x21, 0xfa,
+                            0x5c, 0xfa, 0x08, 0x02, 0xc6, 0xe1, 0x1f, 0xd0, 0xed, 0xbe};
+  EXPECT_THROW(src_repo->GetObject(hash), OSTreeObjectMissing);
+}
+
+TEST(http_repo, bad_connection) {
+  TemporaryDirectory temp_dir;
+  std::string sp = TestUtils::getFreePort();
+
+  TestHelperProcess server_process("tests/sota_tools/treehub_server.py", sp, "3");
+  sleep(3);
+
+  TreehubServer server;
+  server.root_url("http://localhost:" + sp);
+  OSTreeRepo::ptr src_repo = std::make_shared<OSTreeHttpRepo>(&server);
+
+  std::string dp = TestUtils::getFreePort();
+  Json::Value auth;
+  auth["ostree"]["server"] = std::string("https://localhost:") + dp;
+  Utils::writeFile(temp_dir.Path() / "auth.json", auth);
+  TestHelperProcess deploy_server_process("tests/sota_tools/treehub_deploy_server.py", dp);
+  sleep(3);
+
+  boost::filesystem::path filepath = (temp_dir.Path() / "auth.json").string();
+  boost::filesystem::path cert_path = "tests/fake_http_server/client.crt";
+
+  const uint8_t hash[32] = {0x16, 0xef, 0x2f, 0x26, 0x29, 0xdc, 0x92, 0x63, 0xfd, 0xf3, 0xc0,
+                            0xf0, 0x32, 0x56, 0x3a, 0x2d, 0x75, 0x76, 0x23, 0xbb, 0xc1, 0x1c,
+                            0xf9, 0x9d, 0xf2, 0x5c, 0x3c, 0x3f, 0x25, 0x8d, 0xcc, 0xbe};
+  UploadToTreehub(src_repo, ServerCredentials(filepath), OSTreeHash(hash), cert_path.string(), false, 1);
+
+  int result = system(
+      (std::string("diff -r ") + (src_repo->root() / "objects/").string() + " tests/sota_tools/repo/objects/").c_str());
+  EXPECT_EQ(result, 0);
```

Comment:
This looks great. My one suggestion is to force even more failures, like every other try instead of every third.